### ISSUE TITLE
NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE with `@CheckForNull` annotation

### ISSUE THREAD

**Issue by hgschmie** [spotbugs/spotbugs#1778]
This code:

```java
package testcode;

import static java.util.Objects.requireNonNull;

import java.util.function.Function;
import javax.annotation.CheckForNull;

public class SpotbugsTest {

    @CheckForNull
    public static Function<String, String> get() {
        return k -> "nope";
    }

    public static String getMapping(String value) {

        requireNonNull(value);

        Function<String, String> f = get();
        requireNonNull(f);
        return f.apply(value);
    }
}
```

reports `[ERROR] Medium: Possible null pointer dereference in testcode.SpotbugsTest.getMapping(String, String) due to return value of called method [testcode.SpotbugsTest, testcode.SpotbugsTest] Method invoked at SpotbugsTest.java:[line 20]Known null at SpotbugsTest.java:[line 19] NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE`

`Objects#requireNonNull` is supposed to do the null check, but is flagged as dereferencing its parameter.

Removing the `@CheckForNull` annotation makes the spotbugs check pass.

IntelliJ flow analysis shows this correctly:

(with `requireNonNull` commented out):
<img width="607" alt="Screen Shot 2021-10-27 at 20 08 49" src="https://user-images.githubusercontent.com/39495/139179540-a8e568bd-7401-4976-b56c-c74ece523fd5.png">

There is no such warning with the method call.

**Comment by hgschmie** [comment]
pretty pretty ping?

**Comment by Kidlike** [comment]
I hit the same problem (bug?).

I got rid of the warning by replacing `Objects#requireNonNull` with `Optional#ofNullable`, which in the end I liked more, because I can control the exception being thrown.

**Comment by hgschmie** [comment]
I can still reproduce this bug with spotbugs 4.7.0

**Comment by hgschmie** [comment]
another pretty ping. This is now open for 18 months.

**Comment by gtoison** [comment]
Sorry for the belated answer, the project is a bit low on manpower and it is quite complicated as well (it is static flow analysis after all...)

Can you try:
`value = requireNonNull(value);`
Instead of:
`requireNonNull(value);`

I've seen elsewhere that SpotBugs knows that the return of `requireNonNull` is non null, but it is not smart enough that an exception is thrown otherwise

**Comment by malogulko** [comment]
Hi! Thanks for your proposal, but ```value = requireNonNull(value);``` still triggers spot bugs failure.
